File: kotlin/src/atcoder/ABC120/D.kt
package atcoder.ABC120

fun main(args: Array<String>) {
    val (n, m) = readLine()!!.split(" ").map { it.toInt() }
    val list = (1..m).map { readLine()!!.split(" ").map { it.toInt() } }.reversed()

    val ans = mutableListOf<Long>()
    val uf = UnionFind(n)

    // 不便さ
    var cur = (n.toLong() * (n.toLong() - 1)) / 2

    // 逆順にみていく(辺がない状態から辺を足していく)
    for (ab in list) {
        ans.add(cur)

        val a = ab[0] - 1
        val b = ab[1] - 1

        // 同じ連結成分なら不便さは変わらない
        if (uf.isSame(a, b)) {
            continue
        }

        // aとbをつなげると、それぞれが属する連結成分の大きさをかけた数だけ行き来できるルートが増えるので
        // そのぶん不便さが減る
        val sizeA = uf.size(a)
        val sizeB = uf.size(b)
        cur -= sizeA * sizeB

        uf.merge(a, b)  // aとbをつなげる
    }

    ans.reversed().forEach{ println(it) }
}

private class UnionFind(n: Int) {
    val parent = Array(n) { -1 }

    fun root(x: Int): Int {
        return if (parent[x] < 0) {
            x
        } else {
            parent[x] = root(parent[x])
            parent[x]
        }
    }

    fun isSame(x: Int, y: Int): Boolean {
        return root(x) == root(y)
    }

    fun merge(x: Int, y: Int): Boolean {
        var xRoot = root(x)
        var yRoot = root(y)
        if (xRoot == yRoot) {
            return false
        }

        // merge technique
        if (parent[xRoot] > parent[yRoot]) {
            val tmp = yRoot
            yRoot = xRoot
            xRoot = tmp
        }

        parent[xRoot] += parent[yRoot];
        parent[yRoot] = xRoot
        return true
    }

    fun size(x: Int): Int {
        return -parent[root(x)]
    }
}

File: rust/atcoder/ABC120/d.rs
macro_rules! input {
    (source = $s:expr, $($r:tt)*) => {
        let mut iter = $s.split_whitespace();
        let mut next = || { iter.next().unwrap() };
        input_inner!{next, $($r)*}
    };
    ($($r:tt)*) => {
        let stdin = std::io::stdin();
        let mut bytes = std::io::Read::bytes(std::io::BufReader::new(stdin.lock()));
        let mut next = move || -> String{
            bytes
                .by_ref()
                .map(|r|r.unwrap() as char)
                .skip_while(|c|c.is_whitespace())
                .take_while(|c|!c.is_whitespace())
                .collect()
        };
        input_inner!{next, $($r)*}
    };
}

macro_rules! input_inner {
    ($next:expr) => {};
    ($next:expr, ) => {};

    ($next:expr, $var:ident : $t:tt $($r:tt)*) => {
        let $var = read_value!($next, $t);
        input_inner!{$next $($r)*}
    };
}

macro_rules! read_value {
    ($next:expr, ( $($t:tt),* )) => {
        ( $(read_value!($next, $t)),* )
    };

    ($next:expr, [ $t:tt ; $len:expr ]) => {
        (0..$len).map(|_| read_value!($next, $t)).collect::<Vec<_>>()
    };

    ($next:expr, chars) => {
        read_value!($next, String).chars().collect::<Vec<char>>()
    };

    ($next:expr, usize1) => {
        read_value!($next, usize) - 1
    };

    ($next:expr, $t:ty) => {
        $next().parse::<$t>().expect("Parse error")
    };
}

fn main() {
    input! {
        n: usize,
        m: usize,
        v: [(usize1, usize1); m],
    }

    let mut uf = UnionFind::new(n);

    let mut ans = vec![];
    let mut cur: i64 = (n as i64 * (n - 1) as i64) / 2;

    for (a, b) in v.into_iter().rev() {
        ans.push(cur);
        
        if uf.is_same(a, b) {
            continue;
        }

        let size_a = uf.size(a);
        let size_b = uf.size(b);
        cur -= (size_a * size_b) as i64;

        uf.merge(a, b);
    }

    ans.reverse();

    for a in ans {
        println!("{}", a);
    }
}

struct UnionFind {
    parent: Vec<i32>,
}

impl UnionFind {
    fn new(n: usize) -> UnionFind {
        // let mut v = vec![-1, n];
        UnionFind { parent: vec![-1; n] }
    }

    fn root(&mut self, x: usize) -> usize {
        if self.parent[x] < 0 {
            x as usize
        } else {
            let x2 = self.parent[x] as usize;
            self.parent[x] = self.root(x2) as i32;
            self.parent[x] as usize
        }
    }

    fn is_same(&mut self, x: usize, y: usize) -> bool {
        self.root(x) == self.root(y)
    }

    fn merge(&mut self, x: usize, y: usize) -> bool {
        let mut x_root = self.root(x);
        let mut y_root = self.root(y);
        if x_root == y_root {
            return false;
        }

        // merge technique
        if self.parent[x_root] > self.parent[y_root] {
            let tmp = x_root;
            x_root = y_root;
            y_root = tmp;
        }

        self.parent[x_root] += self.parent[y_root];
        self.parent[y_root] = x_root as i32;
        return true;
    }

    fn size(&mut self, x: usize) -> usize {
        let root = self.root(x);
        -self.parent[root] as usize
    }

    fn print(&mut self) {
        println!("{:?}", self.parent);
    }
}